File: bin/pipeline_sender.cpp
/*
 * pipeline_sender - ZMQ PUSH sender with verifiable sequence-numbered messages
 *
 * Message format (compatible with pipeline_validator):
 *   bytes 0-7 : uint64 big-endian sequence number (starts at 0)
 *   bytes 8+  : fill pattern where each byte == (seq_num & 0xFF)
 *
 * Used as N1 in the pipeline test:
 *   pipeline_sender -> zmq_ejfat_bridge -> ejfat_zmq_proxy -> pipeline_validator
 */

#include <zmq.hpp>
#include <boost/program_options.hpp>
#include <iostream>
#include <atomic>
#include <csignal>
#include <thread>
#include <chrono>
#include <cstring>
#include <vector>

namespace po = boost::program_options;

static std::atomic<bool> g_stop{false};

void signalHandler(int) {
    g_stop.store(true);
}

static inline uint64_t to_be64(uint64_t v) {
#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
    return __builtin_bswap64(v);
#else
    return v;
#endif
}

int main(int argc, char* argv[]) {
    po::options_description desc("Pipeline ZMQ PUSH Sender");
    desc.add_options()
        ("help,h",     "Show this help")
        ("endpoint,e", po::value<std::string>()->default_value("tcp://*:5556"),
                       "ZMQ PUSH bind endpoint")
        ("count,n",    po::value<uint64_t>()->default_value(1000),
                       "Number of messages to send (0=unlimited)")
        ("size,s",     po::value<int>()->default_value(4096),
                       "Message size in bytes (min: 8)")
        ("rate,r",     po::value<int>()->default_value(100),
                       "Messages per second (0=unlimited)")
        ("hwm",        po::value<int>()->default_value(10000),
                       "ZMQ send HWM");

    po::variables_map vm;
    try {
        po::store(po::parse_command_line(argc, argv, desc), vm);
        if (vm.count("help")) {
            std::cout << desc << std::endl;
            return 0;
        }
        po::notify(vm);
    } catch (const po::error& e) {
        std::cerr << "ERROR: " << e.what() << std::endl;
        std::cerr << desc << std::endl;
        return 1;
    }

    const int msg_size = vm["size"].as<int>();
    if (msg_size < 8) {
        std::cerr << "ERROR: --size must be >= 8 (sequence number header is 8 bytes)" << std::endl;
        return 1;
    }

    const uint64_t count    = vm["count"].as<uint64_t>();
    const int      rate     = vm["rate"].as<int>();
    const bool     unlimited = (count == 0);

    std::signal(SIGINT,  signalHandler);
    std::signal(SIGTERM, signalHandler);

    zmq::context_t ctx(1);
    zmq::socket_t  sock(ctx, zmq::socket_type::push);
    sock.set(zmq::sockopt::sndhwm, vm["hwm"].as<int>());
    sock.bind(vm["endpoint"].as<std::string>());

    // Pre-allocate reusable send buffer
    std::vector<uint8_t> buf(static_cast<size_t>(msg_size), 0);

    const std::chrono::nanoseconds sleep_ns =
        (rate > 0) ? std::chrono::nanoseconds(1'000'000'000LL / rate)
                   : std::chrono::nanoseconds(0);

    std::cout << "Pipeline sender bound to " << vm["endpoint"].as<std::string>() << std::endl;
    std::cout << "  Message size : " << msg_size << " bytes" << std::endl;
    std::cout << "  Count        : " << (unlimited ? "unlimited" : std::to_string(count)) << std::endl;
    std::cout << "  Rate         : " << (rate == 0 ? "unlimited" : std::to_string(rate) + " msg/s") << std::endl;
    std::cout << "Waiting 1s for downstream connections..." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(1));
    std::cout << "Sending...\n" << std::endl;

    uint64_t sent  = 0;
    auto     start = std::chrono::steady_clock::now();
    auto     next_send = start;

    while (!g_stop.load() && (unlimited || sent < count)) {
        uint64_t seq       = sent;
        uint64_t seq_be    = to_be64(seq);
        uint8_t  fill_byte = static_cast<uint8_t>(seq & 0xFF);

        std::memcpy(buf.data(), &seq_be, 8);
        std::memset(buf.data() + 8, fill_byte, static_cast<size_t>(msg_size - 8));

        zmq::message_t msg(buf.data(), static_cast<size_t>(msg_size));

        auto result = sock.send(msg, zmq::send_flags::dontwait);
        if (!result.has_value()) {
            // Send buffer full — backpressure, yield briefly
            std::this_thread::sleep_for(std::chrono::microseconds(1));
            continue;
        }

        sent++;

        if (sent % 1000 == 0) {
            auto now     = std::chrono::steady_clock::now();
            double elapsed = std::chrono::duration<double>(now - start).count();
            double r     = (elapsed > 0) ? (static_cast<double>(sent) / elapsed) : 0.0;
            std::cout << "Sent " << sent << " | " << static_cast<int>(r) << " msg/s" << std::endl;
        }

        if (rate > 0) {
            next_send += sleep_ns;
            auto now = std::chrono::steady_clock::now();
            if (next_send > now)
                std::this_thread::sleep_until(next_send);
        }
    }

    auto end     = std::chrono::steady_clock::now();
    double elapsed = std::chrono::duration<double>(end - start).count();
    double avg_rate = (elapsed > 0) ? (static_cast<double>(sent) / elapsed) : 0.0;

    std::cout << "\n=== Sender Summary ===" << std::endl;
    std::cout << "Messages sent : " << sent << std::endl;
    std::cout << "Duration      : " << elapsed << "s" << std::endl;
    std::cout << "Average rate  : " << static_cast<int>(avg_rate) << " msg/s" << std::endl;
    std::cout << "Throughput    : "
              << static_cast<double>(sent * static_cast<uint64_t>(msg_size)) / 1e9 / elapsed
              << " GB/s" << std::endl;
    std::cout << "======================" << std::endl;

    sock.close();
    ctx.close();
    return 0;
}

File: scripts/MacOS/pipeline_sender.py
#!/usr/bin/env python3
"""
Pipeline test sender: ZMQ PUSH with verifiable sequence-numbered messages.

Message format:
  bytes 0-7 : uint64 big-endian sequence number (starts at 0)
  bytes 8+  : fill pattern where each byte == (seq_num & 0xFF)

Used as N1 in the pipeline test:
  pipeline_sender -> zmq_ejfat_bridge -> ejfat_zmq_proxy -> pipeline_validator
"""

import struct
import zmq
import time
import argparse
import signal
import sys


def main():
    parser = argparse.ArgumentParser(description="Pipeline test ZMQ PUSH sender")
    parser.add_argument("--endpoint", "-e", default="tcp://*:5556",
                        help="ZMQ PUSH bind endpoint (default: tcp://*:5556)")
    parser.add_argument("--count", "-n", type=int, default=1000,
                        help="Number of messages to send (0=unlimited, default: 1000)")
    parser.add_argument("--size", "-s", type=int, default=4096,
                        help="Message size in bytes (default: 4096, min: 8)")
    parser.add_argument("--rate", "-r", type=int, default=100,
                        help="Messages per second (0=unlimited, default: 100)")
    parser.add_argument("--hwm", type=int, default=10000,
                        help="ZMQ send HWM (default: 10000)")
    args = parser.parse_args()

    if args.size < 8:
        print("ERROR: --size must be >= 8 (sequence number header is 8 bytes)",
              file=sys.stderr)
        sys.exit(1)

    stop = [False]

    def on_signal(sig, frame):
        print("\nShutdown signal received...")
        stop[0] = True

    signal.signal(signal.SIGINT,  on_signal)
    signal.signal(signal.SIGTERM, on_signal)

    ctx = zmq.Context()
    sock = ctx.socket(zmq.PUSH)
    sock.set(zmq.SNDHWM, args.hwm)
    sock.bind(args.endpoint)

    sleep_time = (1.0 / args.rate) if args.rate > 0 else 0
    unlimited = (args.count == 0)

    print(f"Pipeline sender bound to {args.endpoint}")
    print(f"  Message size : {args.size} bytes")
    print(f"  Count        : {'unlimited' if unlimited else args.count}")
    print(f"  Rate         : {'unlimited' if args.rate == 0 else str(args.rate) + ' msg/s'}")
    print("Waiting 1s for downstream connections...")
    time.sleep(1.0)
    print("Sending...\n")

    sent  = 0
    start = time.time()

    try:
        while not stop[0] and (unlimited or sent < args.count):
            seq = sent
            fill_byte = seq & 0xFF
            payload = struct.pack(">Q", seq) + bytes([fill_byte] * (args.size - 8))

            try:
                sock.send(payload, flags=zmq.DONTWAIT)
                sent += 1
            except zmq.Again:
                # Send buffer full — backpressure, yield briefly
                time.sleep(0.001)
                continue

            if sent % 1000 == 0:
                elapsed = time.time() - start
                rate    = sent / elapsed if elapsed > 0 else 0
                print(f"Sent {sent:,} | {rate:,.1f} msg/s")

            if sleep_time > 0:
                time.sleep(sleep_time)

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
    finally:
        sock.close()
        ctx.term()

    elapsed = time.time() - start
    rate    = sent / elapsed if elapsed > 0 else 0
    print(f"\n=== Sender Summary ===")
    print(f"Messages sent : {sent:,}")
    print(f"Duration      : {elapsed:.2f}s")
    print(f"Average rate  : {rate:,.1f} msg/s")
    print(f"======================")


if __name__ == "__main__":
    main()
